platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,670 @@
+have to consider reconnect at all.
+
+### Liveness Under Stress
+
+A node is under stress when it is unable to process events fast enough to keep up with the other nodes in the network.
+The network is under stress when a sufficient number of nodes are under stress. When the network is under stress, all
+nodes together must work to reduce the number of transactions allowed into the network, to give the network a chance
+to recover. The method by which this is done will be covered in another design document related to "dynamic throttling".
+
+The Consensus Layer performs work on events. The more events, the more work. If events are unbounded, then a node
+under stress will eventually run out of memory and crash. If there are no events, then there is virtually no CPU or
+memory used by Consensus. It is therefore critical that the number of events be bounded within any given node,
+even if the node is running slower than other nodes in the network. Each node must maintain _at least_ all non-ancient
+events, and should maintain additional non-expired events (though these could be stored on disk to remove them from
+the memory requirement of the node). In addition, _birth round filtering_ limits the number of very old or far future
+events that a node will accept (see [Birth-Round Filtering](#birth-round-filtering)).
+
+#### CPU Pressure
+
+From a high level, either Execution or Consensus can be the primary bottleneck in handling events.
+
+##### Consensus Bottlenecks
+
+Let us suppose that we have a node, Alice. Perhaps initially Alice is able to receive and process events at the same
+speed as other nodes in the network. Perhaps the network load increases past some point that Alice can handle.  At this
+point, other nodes are receiving, creating, gossiping, and coming to consensus on rounds faster than Alice. Remember:
+
+1. Birth-round filtering limits the number of events received by Alice to coincide with the pace at which Alice is
+   handling consensus.
+2. The Tipset algorithm only creates events when doing so will advance consensus. It relies on events that have already
+   passed through Birth-round filtering.
+
+As the other nodes progress farther than Alice, they begin to send events with newer and newer birth rounds. At some
+point, they get too far ahead and begin delivering events that are too far in the future and fail to pass Alice's
+birth-round filter. This prevents Alice from being overwhelmed by events.
+
+Further, since events are coming more slowly to Alice, she will also create her own events more slowly.
+
+Further, since events are coming more slowly to Alice, her Event Creator will update more slowly. When the tipset
+algorithm is consulted for an "other parent", it will find after a short time that there are no other parents it can
+select that will advance consensus. This will cause it to stop creating events until enough events from other creators
```

Comment:
If Alice is slower than most nodes (she's in the slowest 1/3), then eventually every event coming from the 2/3 will have a future birth round, and will be dropped. The only way tipset will allow here to continue is if she starts getting some of those events. That will only happen as fast as she can handle the work she has (i.e. consensus round has to advance). Each time the consensus round advances, she gets a new batch of (old, to the rest of the network, new to her) events that she can build on.

If Alice is under stress such that her ability towards moving forward consensus is limited to be less than the events/sec limit, then we would see this slowdown. This could happen if for example consensus is super slow (really bad hardware under load) or much more likely if handling transactions is slow, or much more likely even than that, if the node is unable to upload blocks for some period of time and decides it should not advance consensus. That could be an arbitrarily long pause. In that case, I think the description here is accurate, and Alice would stop producing events (after having created what she could) until consensus can move forward.

Reply:
In the case where consensus is more or less *entirely* blocked, I agree that Alice will stop creating events. 

But the implication as currently written seems to be that if Alice gets gradually slower, she will gradually create fewer events. I don't agree with is that Alice would limit event creation simply because she is *slower* than the rest of the network. She needs to be *massively* slower.

Let's say Alice is struggling and can only process events at 50% the rate of the rest of the network. Even though her event creator is receiving new events *slower* due to her stress, even the decreased event rate will likely be enough to supply enough other-parents to create the same number of events as normal. I think the potential for a "virtuous cycle" here is very limited.

Reply:
>Let's say Alice is struggling and can only process events at 50% the rate of the rest of the network. Even though her event creator is receiving new events slower due to her stress, even the decreased event rate will likely be sufficient to supply enough other-parents to create the same number of events as normal.

I agree.

Reply:
I had a conversation with Leemon last week, where we were exploring tying tipset to hashgraph at a deeper level, whereas right now tipset snapshots are independent of hashgraph consensus rounds etc. If we did tie them together in a more fundamental way, we could cause events-per-round to be fixed, such that events-per-second would be emergent based on the speed by which you come to consensus, in which case I think you would see a more gradual slowdown / speedup in the event creator based on your consensus speed. It was a brainstorm idea, needs to be considered more.